### PR TITLE
Adding support for SVG <image> tags

### PR DESCRIPTION
**Pull request by Lucas-C**
This is a contribution towards #537

What's missing before merging:
* clean-up the dependencies: having `SVGImage.render()` calling `fpdf.stream_content_for_image()` is really non-ideal
* implement remaining edge cases: `<image>` without explicit `width`/`height`, `SVGImage.render_debug()`
* more tests

### CODE REVIEW

**Comment by Lucas-C** [comment]
I think this is ready for a review @gmischler & @andersonhc 🙂

**Comment by pnikhil610** [comment]
it would be awesome if it can be merged :+1: 
Been waiting for this feature

File: fpdf/svg.py [hunk 1]
```diff
@@ -845,6 +852,7 @@ def draw_to_page(self, pdf, x=None, y=None, debug_stream=None):
             debug_stream (io.TextIO): the stream to which rendering debug info will be
                 written.
         """
+        self.img_cache = pdf  # Needed to render images
```
Comment (gmischler) [inline fpdf/svg.py:855]: This is extremely confusing to read.

I see that you let FPDF implement ImageCacheI, so it obviously works. But since the image cache is only ever used internally, wouldn't it make more sense to use a singleton ImageCache instance as an attribute? That would make it immediately recognizable what's actually going on, and not result in such weird assignments like here.

File: fpdf/fpdf.py [hunk 2]
```diff
@@ -78,19 +78,21 @@ class Image:
 from .fonts import CoreFont, CORE_FONTS, FontFace, TTFFont
 from .graphics_state import GraphicsStateMixin
 from .html import HTML2FPDF
+from .image_datastructures import (
```
Comment (gmischler) [inline fpdf/fpdf.py:81]: Nice continuation on my (so far rather superficial) start to better encapsulate image handling and unify raster and vector processing.

Your changes will conflict with mine in the upcoming table cell/text region PR, but that can't be helped...

On the other hand, image_parsing.py is not all that big. Is it really worth it to seperate out the data structures?

File: fpdf/fpdf.py [hunk 3]
```diff
@@ -247,11 +249,13 @@ def __init__(
         self.page = 0  # current page number
         self.pages = {}  # array of PDFPage objects starting at index 1
         self.fonts = {}  # map font string keys to an instance of CoreFont or TTFFont
-        self.images = {}  # map image identifiers to dicts describing the raster images
-        self.icc_profiles = {}  # map icc profiles (bytes) to their index (number)
         self.links = {}  # array of Destination objects starting at index 1
         self.embedded_files = []  # array of PDFEmbeddedFile
 
+        # ImageCacheI:
```
Comment (gmischler) [inline fpdf/fpdf.py:255]: This should really be encapsulated in its own instance (see my other comment).

File: fpdf/output.py [hunk 4]
```diff
@@ -952,6 +953,31 @@ def _log_final_sections_sizes(self):
             LOGGER.debug("- %s: %s", label, _sizeof_fmt(section_size))
 
 
+def stream_content_for_image(
```
Comment (gmischler) [inline fpdf/output.py:956]: It looks like this only applies to raster images. If so, then the name should probably reflect that.

File: fpdf/svg.py [hunk 5]
```diff
@@ -978,3 +984,89 @@ def apply_clipping_path(self, stylable, svg_element):
...
             clipping_path_id = re.search(r"url\((\#\w+)\)", clipping_path)
             stylable.clipping_path = self.cross_references[clipping_path_id[1]]
+
+    @force_nodocument
+    def build_image(self, image):
+        if xmlns("xlink", "href") not in image.attrib:
+            raise ValueError("<image> is missing a href attribute")
+        width = float(image.attrib.get("width", 0))
+        height = float(image.attrib.get("height", 0))
+        if "preserveAspectRatio" in image.attrib:
```
Comment (gmischler) [inline fpdf/svg.py:994]: It might be nicer to only emit warnings for some of these cases. The result may not look like the user expected, but at least they get something.
Note that I have no idea how these attributes are actually used in practise, so this suggestion may or may not make sense (to be determined on a case-by-case basis).

**Review by gmischler — CHANGES_REQUESTED** [review]
A very nice and helpful feature overall!

A few architectural questions still to be considered.

File: fpdf/output.py (hunk 4 above)
Comment (Lucas-C) [reply fpdf/output.py:956]: Agreed, I'm going to rename this function to include `_raster_` in it

File: fpdf/fpdf.py (hunk 2 above)
Comment (Lucas-C) [reply fpdf/fpdf.py:81]: > On the other hand, image_parsing.py is not all that big. Is it really worth it to seperate out the data structures?

Yes, this change was needed to avoid circular imports between `fpdf.image_parsing` & `fpdf.fpdf` modules.

The new `fpdf.image_datastructures` module is used by no less than 5 modules: `fpdf.image_parsing`, `fpdf.output`, `fpdf.fpdf`, `fpdf.svg` and `fpdf.text_region`.
On the other hand, `fpdf.image_parsing` is only used by 3 modules: `fpdf.fpdf`, `fpdf.text_region` and `fpdf.svg` (with an ugly function-local import because I couldn't figure a better way to do it, for now).

In the future, maybe we should initiate a `fpdf/images/` sub-folder to gather all image-related modules?

File: fpdf/fpdf.py (hunk 3 above)
Comment (Lucas-C) [reply fpdf/fpdf.py:255]: Alright. I'm going to answer your other comment

File: fpdf/svg.py (hunk 5 above)
Comment (Lucas-C) [reply fpdf/svg.py:994]: Yes, I considered this...

I prefer to be "strict" for now, and not allow those fields because:
* ignoring `transform` attributes will have a huge impact on the final layout of the vector graphics rendered
* we could make this behaviour configurable later one (with a "ignore_unsupported_svg_features" flag maybe?), but it's safer to switch from errors to warnings without bothering users, than the other way round (from warnings to errors)
* with visible errors like this, I hope that the `contributions are welcome!` message will attract contributors 😅

File: fpdf/svg.py (hunk 1 above)
Comment (Lucas-C) [reply fpdf/svg.py:855]: > wouldn't it make more sense to use a singleton ImageCache instance as an attribute?

Yes, you are totally right!

I'll do that right now, thank you for wise feedback 🙂

File: fpdf/fpdf.py (hunk 2 above)
Comment (gmischler) [reply fpdf/fpdf.py:81]: > Yes, this change was needed to avoid circular imports between fpdf.image_parsing & fpdf.fpdf modules.

I think you mean between `fpdf.image_parsing` and `fpdf.svg`, there should not have been a cross-dependency with `fpdf.fpdf`. But I did indeed forget that `fpdf.image_parsing` uses `SVGObject` which makes it tricky.

> In the future, maybe we should initiate a fpdf/images/ sub-folder to gather all image-related modules?

Are you expecting more than two image related source files in the future to justify a subdirectory?
Before we start with subdirectories in general, I'd prefer to establish an overall concept for that (eg. core/utils/graphics/text/layout or something similar). But I really don't think this is urgent, we still have less than 30 source files.

Comment (Lucas-C) [reply fpdf/fpdf.py:81]: > I think you mean between `fpdf.image_parsing` and `fpdf.svg`, there should not have been a cross-dependency with `fpdf.fpdf`. But I did indeed forget that `fpdf.image_parsing` uses `SVGObject` which makes it tricky.

Yes, that's what I meant to write 😅

> Are you expecting more than two image related source files in the future to justify a subdirectory?

I thought about maybe putting `fpdf/svg.py` in it too.

> But I really don't think this is urgent, we still have less than 30 source files.

Agreed

File: fpdf/svg.py (hunk 5 above)
Comment (gmischler) [reply fpdf/svg.py:994]: It will be interesting to see which one of those we're going to be asked about first, since I really have no idea how often and in which ways they are used in practise.

**Review by gmischler — APPROVED** [review]
Looking good.

You don't seem to have implemented the use of images as xref from defs. Is that more complicated to do for images than for the other elements? Otherwise it might be worth adding it.

**Comment by Lucas-C** [comment]
> You don't seem to have implemented the use of images as xref from defs. Is that more complicated to do for images than for the other elements? Otherwise it might be worth adding it.

You mean support for `<image>` with a `href` attribute value being `#some-id`, with `some-id` defined in a `<def>` tag?

**Comment by gmischler** [comment]
> Could you please provide an example?

I couldn't get InkScape to create one directly, but I could modify one:

```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
   xmlns:svg="http://www.w3.org/2000/svg"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:xlink="http://www.w3.org/1999/xlink"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   width="50mm"
   height="50mm"
   viewBox="0 0 100 100"
   version="1.1"
   id="svg8"
   >

  <defs id="defs2" >
     <image
 xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQAQMAAAAlPW0iAAAAAHViVUMe7ynHAAAAAHViU2Nz266J
AAAAA1BMVEUAgACc+aWRAAAACXBIWXMAAABIAAAASABGyWs+AAAACXZwQWcAAAAQAAAAEABcxq3D
AAAADElEQVQI12NgIA0AAAAwAAHHqoWOAAAAAHVlVUMboD9CAAAAAHVlU2N2lLgMAAAAJXRFWHRj
cmVhdGUtZGF0ZQAyMDDULTA0LTA3VDE5OjA2OjE2LTA3OjAwNE1NWwAAACV0RVh0bW9kaWZ5LWRh
dGVjxjAwOC0wNC0wN1QxOTowNjoxNi0wNzowMD/5zc0AAAAASUVORK5CYII=
"
       id="use4545"
       preserveAspectRatio="none"
       height="10"
       width="10" />
  </defs>

  <g
     inkscape:label="Ebene 1"
     inkscape:groupmode="layer"
     id="layer1">

    <use
       x="10"
       y="10"
       xlink:href="#use4545"
       id="use4547"
       width="100%"
       height="100%"
    />
  </g>
</svg>
```

Of course that would be most useful for placing the same image in different locations, which would require supporting `transform`...
defs-image

**Comment by Lucas-C** [comment]
> I couldn't get InkScape to create one directly, but I could modify one:

Thank you!

I added a commit to this PR to support this usage
